lib/ddtrace/configuration/components.rb

```diff
@@ -1,106 +1,153 @@
 require 'ddtrace/diagnostics/health'
 require 'ddtrace/logger'
+require 'ddtrace/runtime/metrics'
 require 'ddtrace/tracer'
+require 'ddtrace/workers/runtime_metrics'
 
 module Datadog
   module Configuration
     # Global components for the trace library.
     # rubocop:disable Metrics/LineLength
     class Components
-      def initialize(settings)
-        # Logger
-        @logger = build_logger(settings)
+      class << self
+        def replace!(old, settings)
+          replacement = new(settings)
+          old.teardown!(replacement)
+          replacement
+        end
 
-        # Tracer
-        @tracer = build_tracer(settings)
+        def build_health_metrics(settings)
+          settings = settings.diagnostics.health_metrics
+          options = { enabled: settings.enabled }
+          options[:statsd] = settings.statsd unless settings.statsd.nil?
 
-        # Runtime metrics
-        build_runtime_metrics(settings)
+          Datadog::Diagnostics::Health::Metrics.new(options)
+        end
 
-        # Health metrics
-        @health_metrics = build_health_metrics(settings)
-      end
+        def build_logger(settings)
+          logger = settings.logger.instance || Datadog::Logger.new(STDOUT)
+          logger.level = settings.diagnostics.debug ? ::Logger::DEBUG : settings.logger.level
 
-      attr_reader \
-        :health_metrics,
-        :logger,
-        :tracer
+          logger
+        end
 
-      def runtime_metrics
-        tracer.writer.runtime_metrics
-      end
+        def build_runtime_metrics(settings)
+          options = { enabled: settings.runtime_metrics.enabled }
+          options[:statsd] = settings.runtime_metrics.statsd unless settings.runtime_metrics.statsd.nil?
+          options[:services] = [settings.service] unless settings.service.nil?
 
-      private
+          Datadog::Runtime::Metrics.new(options)
+        end
 
-      def build_logger(settings)
-        logger = settings.logger.instance || Datadog::Logger.new(STDOUT)
-        logger.level = settings.diagnostics.debug ? ::Logger::DEBUG : settings.logger.level
+        def build_runtime_metrics_worker(settings)
+          # NOTE: Should we just ignore building the worker if its not enabled?
```

Comment:
For the time being, because we still allow dynamic reconfiguration, I think always initializing it is ok.

Reply:
Okay, I'll remove this for now... we'll consider optimizations like this later.

Reply:
I think the comment is actually helpful for future work towards true immutable components: it guides us to where we need to make changes.

I think I saw the comment as a "request for comment" and I was just chiming in on what I think about it.

Reply:
Yeah, that is kind of the spirit of writing it, so this is good.